src/Url.php

```diff
@@ -2,70 +2,298 @@
 
 namespace AmpProject;
 
+use AmpProject\Exception\FailedToParseUrl;
+
 /**
  * Helper class to work with URLs.
  *
+ * @property-read string|null $scheme
+ * @property-read string|null $host
+ * @property-read string|null $port
+ * @property-read string|null $user
+ * @property-read string|null $pass
+ * @property-read string|null $path
+ * @property-read string|null $query
+ * @property-read string|null $fragment
+ *
  * @package ampproject/amp-toolbox
  */
 final class Url
 {
 
+    const SCHEME   = 'scheme';
+    const HOST     = 'host';
+    const PORT     = 'port';
+    const USER     = 'user';
+    const PASS     = 'pass';
+    const PATH     = 'path';
+    const QUERY    = 'query';
+    const FRAGMENT = 'fragment';
+
+    /**
+     * Regular expression pattern used to collapse the current path ('.').
+     *
+     * @var string
+     */
+    const COLLAPSE_CURRENT_PATHS_REGEX_PATTERN = '#/\./#';
+
+    /**
+     * Regular expression pattern used to collapse a relative path ('..').
+     *
+     * @var string
+     */
+    const COLLAPSE_RELATIVE_PATHS_REGEX_PATTERN = '#(?<=/)(?!\.\./)[^/]+/\.\./#';
+
+    /**
+     * Error message to use when the __get() is triggered for an unknown property.
+     *
+     * @var string
+     */
+    const PROPERTY_GETTER_ERROR_MESSAGE = 'Undefined property: AmpProject\\Url::';
+
+    /**
+     * Scheme.
+     *
+     * @var string|null
+     */
+    private $scheme;
+
+    /**
+     * Host.
+     *
+     * @var string|null
+     */
+    private $host;
+
+    /**
+     * Port.
+     *
+     * @var string|null
+     */
+    private $port;
+
+    /**
+     * User.
+     *
+     * @var string|null
+     */
+    private $user;
+
+    /**
+     * Password.
+     *
+     * @var string|null
+     */
+    private $pass;
+
+    /**
+     * Query string.
+     *
+     * @var string|null
+     */
+    private $query;
```

Comment:
Eventually it will be nice to add a way to access/manipulate the query vars as an array.

Reply:
That would be powered by a separate object, so that you'd be able to use it as part of a URL, or stand-alone when dealing with WP query vars, for example.